test/base/Pie.test.hh

```diff
@@ -0,0 +1,163 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Pie.test.hh
+//---------------------------------------------------------------------------//
+
+#include "base/Pie.hh"
+#include "base/Types.hh"
+
+namespace celeritas_test
+{
+using celeritas::MemSpace;
+using celeritas::Ownership;
+
+//---------------------------------------------------------------------------//
+// MOCK PIES
+//---------------------------------------------------------------------------//
+
+struct MockElement
+{
+    int    atomic_number = 0;
+    double atomic_mass;
+};
+
+struct MockMaterial
+{
+    double                           number_density;
+    celeritas::PieSlice<MockElement> elements;
+};
+
+template<Ownership W, MemSpace M>
+struct MockParamsPies
+{
+    //// TYPES ////
+
+    template<class T>
+    using Pie = celeritas::Pie<T, W, M>;
+
+    //// DATA ////
+
+    celeritas::Pie<MockElement, W, M>  elements;
+    celeritas::Pie<MockMaterial, W, M> materials;
+    int                                max_element_components{};
+
+    //// MEMBER FUNCTIONS ////
+
+    //! Whether the object is in a valid state
+    explicit operator bool() const
+    {
+        return !materials.empty() && max_element_components >= 0;
+    }
+
+    //! Assign from another set of pies
+    template<Ownership W2, MemSpace M2>
+    MockParamsPies& operator=(const MockParamsPies<W2, M2>& other)
+    {
+        CELER_EXPECT(other);
+        elements               = other.elements;
+        materials              = other.materials;
+        max_element_components = other.max_element_components;
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Mock-up of a set of states.
+ */
+template<Ownership W, MemSpace M>
+struct MockStatePies
+{
+    //// TYPES ////
+
+    template<class T>
+    using Pie = celeritas::Pie<T, W, M>;
+
+    //// DATA ////
+
+    celeritas::Pie<int, W, M> matid;
+
+    //// MEMBER FUNCTIONS ////
+
+    explicit CELER_FUNCTION operator bool() const { return !matid.empty(); }
+    CELER_FUNCTION celeritas::size_type size() const { return matid.size(); }
+
+    // NOTE: no constructor from MockStatePies<W2, M2> means that calling
+    //  MockStatePies<ref, host> foo = host_pies;
+    // gives an ugly error message
+
+    //! Assign from another set of pies on the host
+    template<Ownership W2, MemSpace M2>
+    MockStatePies& operator=(MockStatePies<W2, M2>& other)
+    {
+        CELER_EXPECT(other);
```

Comment:
Does that imply that 'empty' `MockStatePies` are "illegal" to manipulate?

Reply:
Yes, by design. In none of our existing use cases do we want to take a reference to an incompletely constructed value, or copy incomplete data to device. I don't think we ever want to do that, but the "requirement" is still going to be enforced on a class-by-class basis rather than the fundamental Pie construct.